### PR TITLE
feat(pyramid): support root graph hops limit

### PR DESCRIPTION
**Pull request by dasurax**
## Change Type

- [ ] Bug fix
- [x] New feature
- [ ] Improvement/Refactor
- [ ] Documentation
- [ ] CI/Build/Infra

## Linked Issue

Not applicable. This PR uses `kind/improvement`, matching the HGraph `hops_limit` change in #1620.

## What Changed

- Add `hops_limit` parsing and documentation for Pyramid KNN search.
- Apply the limit only to the root graph while leaving path sub-graphs unlimited.
- Match HGraph behavior by ignoring `hops_limit <= ef_search` with a warning.
- Reject negative values and values exceeding `UINT32_MAX` instead of narrowing them.
- Add parameter parsing coverage and functional assertions based on reported search hops.

## Test Evidence

- [ ] `make fmt`
- [ ] `make lint`
- [ ] `make test`
- [ ] `make cov`, run tests, and collect coverage
- [x] Other (described below)

Test details:

```text
clang-format 15.0.7 --dry-run --Werror <changed C++ files>
  passed

cmake --build build --target functests --parallel 4
build/tests/functests "Pyramid No Path Test" -s
  All tests passed (40282 assertions in 1 test case)
  unlimited hops=14, limited hops=2, ignored-limit hops=14

cmake --build build --target unittests --parallel 4
build/tests/unittests "Pyramid parses hops limit search parameter"
  All tests passed (4 assertions in 1 test case)

git diff --check
  passed
```

## Compatibility Impact

- API/ABI compatibility: additive public parameter constant; no breaking change
- Behavior changes: optional Pyramid KNN parameter; default behavior remains unlimited

## Performance and Concurrency Impact

- Performance impact: configured root-graph searches may stop earlier; default behavior is unchanged
- Concurrency/thread-safety impact: none

## Documentation Impact

- [ ] No docs update needed
- [x] Updated docs:
  - [ ] `README.md`
  - [ ] `DEVELOPMENT.md`
  - [ ] `CONTRIBUTING.md`
  - [x] Other: English and Chinese Pyramid index documentation

## Risk and Rollback

- Risk level: low
- Rollback plan: revert the commits in this PR

## Checklist

- [x] Linked issue is not required for `kind/improvement`
- [x] I have added/updated tests for new behavior or bug fixes
- [x] I have considered API compatibility impact
- [x] I have updated docs if behavior/workflow changed
- [x] My commit messages follow project conventions (Conventional Commits, optional `[skip ci]` prefix)

### CODE REVIEW

File: src/algorithm/pyramid/pyramid.cpp [hunk 1]
```diff
@@ -386,6 +386,19 @@ Pyramid::KnnSearch(const DatasetPtr& query,
     search_param.enable_rabitq_one_bit_search = parsed_param.has_rabitq_one_bit_search
                                                     ? parsed_param.rabitq_one_bit_search
                                                     : default_rabitq_one_bit_search_;
+
+    // Keep the same contract as HGraph: a useful hop cap must exceed ef_search.
+    if (static_cast<uint64_t>(parsed_param.hops_limit) <= parsed_param.ef_search) {
+        search_param.hops_limit = std::numeric_limits<uint32_t>::max();
+        if (parsed_param.hops_limit != std::numeric_limits<uint32_t>::max()) {
+            logger::warn(
```
Comment (LHT129) [inline src/algorithm/pyramid/pyramid.cpp:394]: [suggestion] `hops_limit` is only applied in `KnnSearch` but not in `RangeSearch`. Both search paths call `search_node` which respects `InnerSearchParam.hops_limit`, so `RangeSearch` silently ignores the user-provided `hops_limit` parameter.

Consider either applying the same hops_limit validation logic in `RangeSearch`, or explicitly documenting that `hops_limit` is KNN-only.

`KnnSearch` has the hops_limit block (pyramid.cpp:389-401):
```cpp
if (static_cast<uint64_t>(parsed_param.hops_limit) <= parsed_param.ef_search) {
    search_param.hops_limit = std::numeric_limits<uint32_t>::max();
    ...
} else {
    search_param.hops_limit = parsed_param.hops_limit;
}
```

`RangeSearch` (pyramid.cpp:433+) does not have this block.

File: src/algorithm/pyramid/pyramid_zparameters.cpp [hunk 2]
```diff
@@ -326,6 +326,9 @@ PyramidSearchParameters::FromJson(const std::string& json_string) {
         obj.subindex_ef_search =
             params[INDEX_PYRAMID][PYRAMID_PARAMETER_SUBINDEX_EF_SEARCH].GetInt();
     }
```
Comment (LHT129) [inline src/algorithm/pyramid/pyramid_zparameters.cpp:328]: [suggestion] `GetInt()` returns `int64_t`, but `hops_limit` is `uint32_t`. If a user passes a negative value or a value exceeding `UINT32_MAX`, the implicit narrowing conversion produces undefined behavior.

Consider adding a validation check, e.g.:
```cpp
int64_t raw = params[INDEX_PYRAMID][PYRAMID_PARAMETER_HOPS_LIMIT].GetInt();
CHECK_ARGUMENT(raw >= 0 && raw <= std::numeric_limits<uint32_t>::max(),
               "hops_limit must be in [0, 4294967295]");
obj.hops_limit = static_cast<uint32_t>(raw);
```

File: src/algorithm/pyramid/pyramid.cpp [hunk 3]
```diff
@@ -1547,6 +1560,9 @@ Pyramid::search_node(const IndexNode* node,
     } else if (node->status_ == IndexNode::Status::GRAPH) {
         InnerSearchParam modified_param = search_param;
         modified_param.ep = node->entry_point_;
+        if (node->level_ != 0) {
+            modified_param.hops_limit = std::numeric_limits<uint32_t>::max();
+        }
         if (node->level_ != 0 && search_param.search_mode == KNN_SEARCH) {
             modified_param.ef =
```
Comment (LHT129) [inline src/algorithm/pyramid/pyramid.cpp]: [suggestion] The condition `node->level_ != 0` is checked twice in adjacent `if` statements (lines 1563 and 1566). These two blocks could be merged into a single `if (node->level_ != 0)` block to avoid the redundant check and make the intent clearer:

```cpp
if (node->level_ != 0) {
    modified_param.hops_limit = std::numeric_limits<uint32_t>::max();
    if (search_param.search_mode == KNN_SEARCH) {
        modified_param.ef = std::min(modified_param.ef,
                         get_suitable_ef_search(search_param.topk, node->graph_->TotalCount(), subindex_ef_search));
    }
}
```

**Review by LHT129 — COMMENTED** [review]
Thanks for this PR. The implementation is clean and follows the HGraph `hops_limit` contract well. I left two inline suggestions:

1. **RangeSearch coverage**: `hops_limit` is only wired into `KnnSearch` but not `RangeSearch`. If this is intentional, please document the limitation; otherwise consider adding the same logic to `RangeSearch`.

2. **Minor refactor**: The two adjacent `if (node->level_ != 0)` blocks in `search_node` could be merged for clarity.

The tests cover parsing defaults, explicit values, and functional hop-count assertions — good coverage. Docs are updated in both English and Chinese. No blocking issues.

File: src/algorithm/pyramid/pyramid.cpp (hunk 1 above)
Comment (dasurax) [reply src/algorithm/pyramid/pyramid.cpp:394]: Thanks. HGraph also applies `hops_limit` only in its KNN branch; its RangeSearch path leaves the default unlimited value. This PR intentionally keeps Pyramid aligned with that behavior, and the Pyramid parameter documentation already describes `hops_limit` as a root-graph KNN limit, so no RangeSearch change is needed.

File: src/algorithm/pyramid/pyramid.cpp (hunk 3 above)
Comment (dasurax) [reply src/algorithm/pyramid/pyramid.cpp]: Updated in 2e946ec0. The non-root hop reset and subindex KNN ef adjustment now share one `node->level_ != 0` block.

File: src/algorithm/pyramid/pyramid_zparameters.cpp (hunk 2 above)
Comment (dasurax) [reply src/algorithm/pyramid/pyramid_zparameters.cpp:328]: Updated in 2e946ec0. Pyramid now validates the parsed value is in `[0, UINT32_MAX]` before the explicit cast, with tests for `-1` and `4294967296`.

**Review by LHT129 — COMMENTED** [review]
Thanks for this PR. The implementation looks clean and well-tested. I reviewed all changed files and here is my assessment:

**What this PR does well:**
- Clean parameter parsing with proper range validation (`CHECK_ARGUMENT` for `[0, UINT32_MAX]`)
- Correctly limits hops only on the root graph (level 0), leaving sub-graphs unlimited
- Matches HGraph contract: ignores `hops_limit <= ef_search` with a warning
- Good test coverage: unit tests for parameter parsing (including negative and overflow rejection), functional tests verifying hop statistics (`unlimited > limited`, `limited <= 2`, `ignored == unlimited`)
- Documentation updated in both English and Chinese

**Existing review comments to address:**
- @LHT129s [comment about RangeSearch](https://github.com/antgroup/vsag/pull/2642#discussion_r3735548027): `hops_limit` is only applied in `KnnSearch` but not `RangeSearch`. Consider either adding it to `RangeSearch` or explicitly documenting it as KNN-only.
- @LHT129s [comment about redundant check](https://github.com/antgroup/vsag/pull/2642#discussion_r3735568772): the two `node->level_ != 0` checks in `search_node` could be merged for clarity.

The validation concern @LHT129 raised about `GetInt()` narrowing has been addressed in the latest commit `2e946ec` with the `CHECK_ARGUMENT` guard — looks good now.

No new issues found from my side. The PR is in good shape pending resolution of the two open review threads.

**Review by LHT129 — COMMENTED** [review]
Reviewed the latest commit 2e946ec0. All previously raised issues have been addressed:

- Parameter validation for `hops_limit` now checks the `[0, UINT32_MAX]` range with `CHECK_ARGUMENT` and includes corresponding unit tests.
- The redundant `node->level_ != 0` check in `search_node` has been consolidated into a single block.
- The `hops_limit` is correctly applied only to the root graph (level 0) while path sub-graphs remain unlimited, matching the documented behavior and HGraph contract.

No new issues found. The implementation is consistent with the existing HGraph `hops_limit` pattern, has adequate test coverage (parameter parsing unit tests + functional integration tests with hop statistics assertions), and the documentation is updated in both English and Chinese.

File: src/algorithm/pyramid/pyramid.cpp [hunk 4]
```diff
@@ -386,6 +386,19 @@ Pyramid::KnnSearch(const DatasetPtr& query,
     search_param.enable_rabitq_one_bit_search = parsed_param.has_rabitq_one_bit_search
                                                     ? parsed_param.rabitq_one_bit_search
                                                     : default_rabitq_one_bit_search_;
+
+    // Keep the same contract as HGraph: a useful hop cap must exceed ef_search.
+    if (static_cast<uint64_t>(parsed_param.hops_limit) <= parsed_param.ef_search) {
+        search_param.hops_limit = std::numeric_limits<uint32_t>::max();
+        if (parsed_param.hops_limit != std::numeric_limits<uint32_t>::max()) {
```
Comment (LHT129) [inline src/algorithm/pyramid/pyramid.cpp:393]: [suggestion] The comparison `static_cast<uint64_t>(parsed_param.hops_limit) <= parsed_param.ef_search` treats an explicit `hops_limit=UINT32_MAX` (4294967295) the same as the default (unset) value. If a user explicitly sets `hops_limit` to `UINT32_MAX`, it will be silently ignored when `ef_search >= UINT32_MAX`, which is correct but the user may expect a warning. Consider distinguishing between "user did not set hops_limit" and "user explicitly set hops_limit=UINT32_MAX" if that distinction matters.

The inner check `parsed_param.hops_limit != std::numeric_limits<uint32_t>::max()` correctly suppresses the warning for the default case, so this is a minor edge case.

File: tests/test_pyramid.cpp [hunk 5]
```diff
@@ -647,6 +669,21 @@ TEST_CASE_PERSISTENT_FIXTURE(fixtures::PyramidTestIndex,
...
+                TestKnnSearch(index, dataset, search_param_invalid_hops_limit, 0.94, true);
+
+                auto query = fixtures::get_one_query(dataset->query_, 0);
+                const auto unlimited_hops = SearchAndGetHops(
+                    index, query, dataset->top_k, search_param_without_hops_limit_for_statistics);
+                const auto limited_hops = SearchAndGetHops(
+                    index, query, dataset->top_k, search_param_for_hops_statistics);
+                const auto ignored_hops = SearchAndGetHops(
+                    index, query, dataset->top_k, search_param_ignored_hops_limit_for_statistics);
+                REQUIRE(unlimited_hops > limited_hops);
```
Comment (LHT129) [inline tests/test_pyramid.cpp:683]: [suggestion] The assertion `REQUIRE(ignored_hops == unlimited_hops)` relies on the HNSW graph traversal producing the exact same number of hops for two searches with the same `ef_search=1`. HNSW traversal is deterministic given the same graph state, but this test would break if any internal change affects hop counting (e.g., entry point selection, tie-breaking). Consider using `REQUIRE(ignored_hops >= unlimited_hops)` or a looser bound to make the test more robust against internal changes that do not affect correctness.

**Review by wxyucs — APPROVED** [review]
lgtm
